docs/developer/editors.md

```diff
@@ -0,0 +1,175 @@
+```
+
+Here is an [example](https://github.com/atomist/automation-client-samples-ts/blob/1da17e847b8e4a55ec246dfac351334ca49f3e71/test/commands/editor/addMyFileTest.ts) of a mocha test for the addMyFile project editor:
+
+```typescript
+it("adds the file", (done) => {
+    const p = InMemoryProject.of({ path: "existingFile.txt", content: "Hi\n" });
+    addMyFile(p).then(r => {
+            assert(p.findFileSync("myFile.txt")getContentSync().includes("I was here");
+        }).then(() => done(), done);
+});
+```
+
+## Working with Projects
+
+The [Project](https://atomist.github.io/automation-client-ts/interfaces/_project_project_.project.html) abstraction aims to provide both synchronous and asynchronous ways of changing projects. The synchronous methods simplify testing,
+while the asynchronous ones will give better performance in your automations.
+
+If you want to modify the filesystem directly in your ProjectEditor function, do what you like; find the underlying directory in `project.baseDir`. 
+
+### Accessing files
+
+If you know the path to the file you want to change, get to it with `findFile`, and dig in it with `getContents`:
+
+```typescript
+import { HandlerContext } from "@atomist/automation-client";
+import { SimpleProjectEditor } from "@atomist/automation-client/operations/edit/projectEditor";
+import { Project } from "@atomist/automation-client/project/Project";
+
+export const changeMyFile: SimpleProjectEditor = (p: Project, ctx: HandlerContext) =>
+        p.findFile("myFile.txt")
+            .then(f => f.getContent()
+                .then(content => f.setContent(
+                    arbitraryManipulation(content))))
+            .then(() => p);
+}
+
+function arbitraryManipulation(oldContent: string): string {
+    return oldContent + " because I said so.\n";
+}
+```
+
+There are many more methods on Project to help manipulate one or more files.
+
+## Changing code
+
+How do we change code in code? There are the tools you're used to, and then the tools unique to Atomist.
+
+### Search and replace
+
+Often, a simple search and replace is enough. Perhaps you want to change a version of a dependency. If you know the file you want to change, and the change you want to make to it. Find the file, and then use `replace` -- it's the same as JavaScript's replace.
+
+The following ProjectEditor finds package.json, finds a declaration about library-of-interest, and replaces the version with `3.4.6`.
+
+```typescript
+import Project from ...
+import File from ...
+
+function upgradeLibrary(project: Project): Promise<Project> {
+    return project.findFile("package.json").then((file: File) => 
+       file.replace(/"library-of-interest": ".*"/, 
+                    `"library-of-interest": "3.4.6"`);
+    )
+}
+```
+
+Find a test for this [here in this place I haven't created yet]
+
+// tip: Please import File from "@atomist/automation-client/project/File", because TypeScript has a built-in File type that's different.
```

Comment:
Do you want

```
!!! tip ""
Please import File from "@atomist/automation-client/project/File", because TypeScript has a built-in File type that's different.
```